### PR TITLE
made sync pipecat observer for dograh

### PR DESCRIPTION
**Pull request by shivamgcodes**
# Pull Request

## Description

Please include a summary of the changes and the related issue. Please also include relevant motivation and context.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Performance improvement
- [ ] Code refactoring
- [ ] Test improvements

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.

- [ ] Unit tests
- [ ] Integration tests
- [ ] Manual testing
- [ ] Performance testing

## Test Configuration

* Python version:
* Operating System:
* Dependencies:

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

## Additional Notes

Add any other context about the pull request here.

## Summary by CodeRabbit

* **New Features**
  * Added synchronous Pipecat task attachment for observer, speech-to-text, pipeline completion, and audio handler setup.
  * Audio recording now starts separately, giving hosts control over when recording begins.
* **Documentation**
  * Updated the changelog and examples to explain the integration behavior and improve formatted output.
* **Tests**
  * Expanded coverage for synchronous attachment, recording control, disabled audio capture, and asynchronous delegation.

<h3>Greptile Summary</h3>

This PR introduces a public synchronous `attach_to_task_sync` method on `NoveumTraceObserver` for hosts (like dograh) that wire the observer outside an event loop. The existing async `attach_to_task` is refactored to delegate all synchronous wiring to the new method and only await the single genuinely-async step — starting `AudioBufferProcessor` recording.

- `attach_to_task` now calls `attach_to_task_sync` (registers observers, safety-net handler, STT detection, and the `on_audio_data` handler) then separately awaits `_ensure_audio_buffer_recording`, cleanly separating sync and async responsibilities.
- `_attach_audio_buffer_from_pipeline` (async) is replaced by `_attach_audio_buffer_handler_sync` (sync), which wires the handler but deliberately omits `start_recording()` — the host calls it directly when using the sync path, or `attach_to_task` calls `_ensure_audio_buffer_recording` in the async path.
- Three new test cases cover: the sync path not starting recording, audio disabled in sync mode, and the async path delegating to sync then starting recording.

<h3>Confidence Score: 4/5</h3>

Safe to merge; the core sync/async split is correct and well-tested, with one small omission in the ABP swap path that matters only in an already-broken teardown scenario.

The refactoring is structurally sound and the new test coverage directly validates the sync path, the delegation chain, and the recording-not-started contract. The only gap is that `_abp_is_recording` is not reset when `_attach_audio_buffer_handler_sync` swaps to a fresh `AudioBufferProcessor`; in the fallback branch of `_ensure_audio_buffer_recording` (processors that don't expose `_recording`), a stale `True` value would suppress `start_recording()` on the new processor. This only triggers if `_finish_conversation` was never called before the swap, which the safety-net handler is specifically designed to prevent — so the issue is latent rather than immediately reachable.

**Files Needing Attention:** src/noveum_trace/integrations/pipecat/pipecat_observer.py — specifically the ABP swap block in `_attach_audio_buffer_handler_sync` and the constructor comment on `_abp_is_recording`.

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| src/noveum_trace/integrations/pipecat/pipecat_observer.py | Core refactor: `attach_to_task` delegates to new public `attach_to_task_sync`; audio handler wiring moved to sync `_attach_audio_buffer_handler_sync` with `start_recording()` deferred to `_ensure_audio_buffer_recording`. `_abp_is_recording` is not reset on ABP swap, a latent edge-case issue. |
| tests/integration/pipecat/test_pipecat_observer.py | Tests updated to reflect new public API: patches changed from `_attach_audio_buffer_from_pipeline` to `_ensure_audio_buffer_recording`, and three new test cases added for `attach_to_task_sync` (sync wiring without recording, audio disabled path, and delegation-to-sync verification). |
| tests/integration/pipecat/test_observer_lifecycle_behavior.py | Removed now-unnecessary `AsyncMock` patches for the deleted `_attach_audio_buffer_from_pipeline` method in two lifecycle tests; both use `record_audio=False` so no audio wiring occurs. |
| src/noveum_trace/integrations/pipecat/tracer.py | Single-line comment update: renamed `_attach_audio_buffer_from_pipeline` reference to `_attach_audio_buffer_handler_sync` to match the refactored method name. |
| CHANGELOG.md | Added entry for version 1.5.21 documenting the new synchronous `attach_to_task_sync` API and the changed behaviour of `attach_to_task` (delegating to the sync path). |
| docs/examples/agent_cleanup_example.py | Cosmetic formatting only: multi-line string literals inside `print()` calls reformatted for PEP 8 compliance. |

</details>

<details open><summary><h3>Sequence Diagram</h3></summary>

```mermaid
sequenceDiagram
    participant Host
    participant attach_to_task
    participant attach_to_task_sync
    participant _attach_audio_buffer_handler_sync
    participant _ensure_audio_buffer_recording
    participant ABP as AudioBufferProcessor

    Note over Host: Async host (event loop available)
    Host->>attach_to_task: await attach_to_task(task)
    attach_to_task->>attach_to_task_sync: attach_to_task_sync(task)
    attach_to_task_sync->>attach_to_task_sync: register TTO/LTO observers
    attach_to_task_sync->>attach_to_task_sync: _register_finish_safety_net_handler
    attach_to_task_sync->>attach_to_task_sync: _detect_pipeline_has_stt
    attach_to_task_sync->>_attach_audio_buffer_handler_sync: _attach_audio_buffer_handler_sync(task)
    _attach_audio_buffer_handler_sync->>ABP: add_event_handler(on_audio_data)
    _attach_audio_buffer_handler_sync-->>attach_to_task_sync: done (no recording started)
    attach_to_task_sync-->>attach_to_task: done
    attach_to_task->>_ensure_audio_buffer_recording: await _ensure_audio_buffer_recording()
    _ensure_audio_buffer_recording->>ABP: await start_recording()
    ABP-->>_ensure_audio_buffer_recording: ok
    _ensure_audio_buffer_recording-->>attach_to_task: done
    attach_to_task-->>Host: done

    Note over Host: Sync host (no event loop)
    Host->>attach_to_task_sync: attach_to_task_sync(task)
    attach_to_task_sync->>_attach_audio_buffer_handler_sync: _attach_audio_buffer_handler_sync(task)
    _attach_audio_buffer_handler_sync->>ABP: add_event_handler(on_audio_data)
    _attach_audio_buffer_handler_sync-->>attach_to_task_sync: done
    attach_to_task_sync-->>Host: done (recording NOT started)
    Host->>ABP: await ABP.start_recording()
```
</details>

<details><summary><h3>Comments Outside Diff (2)</h3></summary>

1. `src/noveum_trace/integrations/pipecat/pipecat_observer.py`, line 801-805 ([link](https://github.com/noveum/noveum-trace/blob/e9820611da7254289ff3e268f3375e31b7d63334/src/noveum_trace/integrations/pipecat/pipecat_observer.py#L801-L805)) 

   <a href="#">P2</a> `_abp_is_recording` is not reset when swapping to a new `AudioBufferProcessor`. `_ensure_audio_buffer_recording` falls back to `_abp_is_recording` only when the processor does not expose `_recording`; if that flag is stale-`True` from the previous ABP, `start_recording()` will be silently skipped for the newly attached processor, preventing any conversation audio from being captured for that session.

   

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: src/noveum_trace/integrations/pipecat/pipecat_observer.py
   Line: 801-805

   Comment:
   `_abp_is_recording` is not reset when swapping to a new `AudioBufferProcessor`. `_ensure_audio_buffer_recording` falls back to `_abp_is_recording` only when the processor does not expose `_recording`; if that flag is stale-`True` from the previous ABP, `start_recording()` will be silently skipped for the newly attached processor, preventing any conversation audio from being captured for that session.

   

   ---

   For each issue above, determine whether it is valid and should be fixed. If so, fix it directly.
   `````
   </details>

   <a href="https://app.greptile.com/ide/claude-code?prompt=This%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20src%2Fnoveum_trace%2Fintegrations%2Fpipecat%2Fpipecat_observer.py%0ALine%3A%20801-805%0A%0AComment%3A%0A%60_abp_is_recording%60%20is%20not%20reset%20when%20swapping%20to%20a%20new%20%60AudioBufferProcessor%60.%20%60_ensure_audio_buffer_recording%60%20falls%20back%20to%20%60_abp_is_recording%60%20only%20when%20the%20processor%20does%20not%20expose%20%60_recording%60%3B%20if%20that%20flag%20is%20stale-%60True%60%20from%20the%20previous%20ABP%2C%20%60start_recording%28%29%60%20will%20be%20silently%20skipped%20for%20the%20newly%20attached%20processor%2C%20preventing%20any%20conversation%20audio%20from%20being%20captured%20for%20that%20session.%0A%0A%60%60%60suggestion%0A%20%20%20%20%20%20%20%20prev_proc%20%3D%20self._audio_buffer_processor%0A%20%20%20%20%20%20%20%20self._audio_buffer_processor%20%3D%20found_proc%0A%20%20%20%20%20%20%20%20self._abp_is_recording%20%3D%20False%20%20%23%20new%20ABP%20%E2%80%94%20recording%20must%20be%20started%20fresh%0A%20%20%20%20%20%20%20%20self._conversation_audio_chunks%20%3D%20%5B%5D%0A%20%20%20%20%20%20%20%20self._conversation_audio_sample_rate%20%3D%20None%0A%20%20%20%20%20%20%20%20self._conversation_audio_num_channels%20%3D%20None%0A%60%60%60%0A%0A---%0A%0AFor%20each%20issue%20above%2C%20determine%20whether%20it%20is%20valid%20and%20should%20be%20fixed.%20If%20so%2C%20fix%20it%20directly.&repo=noveum%2Fnoveum-trace&pr=78&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaudeDark.svg?v=6"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaude.svg?v=6">Fix in Claude Code</picture></a>

2. `src/noveum_trace/integrations/pipecat/pipecat_observer.py`, line 364-366 ([link](https://github.com/noveum/noveum-trace/blob/e9820611da7254289ff3e268f3375e31b7d63334/src/noveum_trace/integrations/pipecat/pipecat_observer.py#L364-L366)) 

   <a href="#">P2</a> The constructor comment says "track internally, don't probe private `_recording`", but `_ensure_audio_buffer_recording` actively prefers Pipecat's `_recording` attribute when present. The comment contradicts the implementation and may mislead anyone trying to understand the fallback ordering.

   

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: src/noveum_trace/integrations/pipecat/pipecat_observer.py
   Line: 364-366

   Comment:
   The constructor comment says "track internally, don't probe private `_recording`", but `_ensure_audio_buffer_recording` actively prefers Pipecat's `_recording` attribute when present. The comment contradicts the implementation and may mislead anyone trying to understand the fallback ordering.

   

   ---

   For each issue above, determine whether it is valid and should be fixed. If so, fix it directly.
   `````
   </details>

   Note: If this suggestion doesn't match your team's coding style, reply to this and let me know. I'll remember it for next time!

   <a href="https://app.greptile.com/ide/claude-code?prompt=This%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20src%2Fnoveum_trace%2Fintegrations%2Fpipecat%2Fpipecat_observer.py%0ALine%3A%20364-366%0A%0AComment%3A%0AThe%20constructor%20comment%20says%20%22track%20internally%2C%20don't%20probe%20private%20%60_recording%60%22%2C%20but%20%60_ensure_audio_buffer_recording%60%20actively%20prefers%20Pipecat's%20%60_recording%60%20attribute%20when%20present.%20The%20comment%20contradicts%20the%20implementation%20and%20may%20mislead%20anyone%20trying%20to%20understand%20the%20fallback%20ordering.%0A%0A%60%60%60suggestion%0A%20%20%20%20%20%20%20%20self._abp_is_recording%3A%20bool%20%3D%20%28%0A%20%20%20%20%20%20%20%20%20%20%20%20False%20%20%23%20fallback%20when%20Pipecat's%20private%20_recording%20is%20absent%0A%20%20%20%20%20%20%20%20%29%0A%60%60%60%0A%0A---%0A%0AFor%20each%20issue%20above%2C%20determine%20whether%20it%20is%20valid%20and%20should%20be%20fixed.%20If%20so%2C%20fix%20it%20directly.&repo=noveum%2Fnoveum-trace&pr=78&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaudeDark.svg?v=6"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaude.svg?v=6">Fix in Claude Code</picture></a>

</details>

<a href="https://app.greptile.com/ide/claude-code?prompt=%23%23%23%20Issue%201%0Asrc%2Fnoveum_trace%2Fintegrations%2Fpipecat%2Fpipecat_observer.py%3A801-805%0A%60_abp_is_recording%60%20is%20not%20reset%20when%20swapping%20to%20a%20new%20%60AudioBufferProcessor%60.%20%60_ensure_audio_buffer_recording%60%20falls%20back%20to%20%60_abp_is_recording%60%20only%20when%20the%20processor%20does%20not%20expose%20%60_recording%60%3B%20if%20that%20flag%20is%20stale-%60True%60%20from%20the%20previous%20ABP%2C%20%60start_recording%28%29%60%20will%20be%20silently%20skipped%20for%20the%20newly%20attached%20processor%2C%20preventing%20any%20conversation%20audio%20from%20being%20captured%20for%20that%20session.%0A%0A%60%60%60suggestion%0A%20%20%20%20%20%20%20%20prev_proc%20%3D%20self._audio_buffer_processor%0A%20%20%20%20%20%20%20%20self._audio_buffer_processor%20%3D%20found_proc%0A%20%20%20%20%20%20%20%20self._abp_is_recording%20%3D%20False%20%20%23%20new%20ABP%20%E2%80%94%20recording%20must%20be%20started%20fresh%0A%20%20%20%20%20%20%20%20self._conversation_audio_chunks%20%3D%20%5B%5D%0A%20%20%20%20%20%20%20%20self._conversation_audio_sample_rate%20%3D%20None%0A%20%20%20%20%20%20%20%20self._conversation_audio_num_channels%20%3D%20None%0A%60%60%60%0A%0A%23%23%23%20Issue%202%0Asrc%2Fnoveum_trace%2Fintegrations%2Fpipecat%2Fpipecat_observer.py%3A364-366%0AThe%20constructor%20comment%20says%20%22track%20internally%2C%20don't%20probe%20private%20%60_recording%60%22%2C%20but%20%60_ensure_audio_buffer_recording%60%20actively%20prefers%20Pipecat's%20%60_recording%60%20attribute%20when%20present.%20The%20comment%20contradicts%20the%20implementation%20and%20may%20mislead%20anyone%20trying%20to%20understand%20the%20fallback%20ordering.%0A%0A%60%60%60suggestion%0A%20%20%20%20%20%20%20%20self._abp_is_recording%3A%20bool%20%3D%20%28%0A%20%20%20%20%20%20%20%20%20%20%20%20False%20%20%23%20fallback%20when%20Pipecat's%20private%20_recording%20is%20absent%0A%20%20%20%20%20%20%20%20%29%0A%60%60%60%0A%0A---%0A%0AFor%20each%20issue%20above%2C%20determine%20whether%20it%20is%20valid%20and%20should%20be%20fixed.%20If%20so%2C%20fix%20it%20directly.&repo=noveum%2Fnoveum-trace&pr=78&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaudeDark.svg?v=6"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaude.svg?v=6">Fix All in Claude Code</picture></a>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
### Issue 1
src/noveum_trace/integrations/pipecat/pipecat_observer.py:801-805
`_abp_is_recording` is not reset when swapping to a new `AudioBufferProcessor`. `_ensure_audio_buffer_recording` falls back to `_abp_is_recording` only when the processor does not expose `_recording`; if that flag is stale-`True` from the previous ABP, `start_recording()` will be silently skipped for the newly attached processor, preventing any conversation audio from being captured for that session.

```suggestion
        prev_proc = self._audio_buffer_processor
        self._audio_buffer_processor = found_proc
        self._abp_is_recording = False  # new ABP — recording must be started fresh
        self._conversation_audio_chunks = []
        self._conversation_audio_sample_rate = None
        self._conversation_audio_num_channels = None
```

### Issue 2
src/noveum_trace/integrations/pipecat/pipecat_observer.py:364-366
The constructor comment says "track internally, don't probe private `_recording`", but `_ensure_audio_buffer_recording` actively prefers Pipecat's `_recording` attribute when present. The comment contradicts the implementation and may mislead anyone trying to understand the fallback ordering.

```suggestion
        self._abp_is_recording: bool = (
            False  # fallback when Pipecat's private _recording is absent
        )
```

---

For each issue above, determine whether it is valid and should be fixed. If so, fix it directly.
`````

</details>

<sub>Reviews (1): Last reviewed commit: ["refactor: remove dead \_attach\_audio\_buff..."](https://github.com/noveum/noveum-trace/commit/e9820611da7254289ff3e268f3375e31b7d63334) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=46727533)</sub>

### CODE REVIEW

File: src/noveum_trace/integrations/pipecat/pipecat_observer.py [hunk 1]
```diff
@@ -752,31 +809,57 @@ async def _attach_audio_buffer_from_pipeline(self, task: Any) -> None:
...
+            self._abp_is_recording = True
+            logger.debug(
+                "Attached to AudioBufferProcessor; start_recording() completed"
+            )
+        except Exception as e:
+            logger.warning(
+                "Failed to start AudioBufferProcessor recording: %s",
+                e,
+                exc_info=True,
+            )
```
Comment (Vidushi2709) [inline src/noveum_trace/integrations/pipecat/pipecat_observer.py]: Right now we have the same logic in two places.

Both `attach_to_task()` and `_attach_audio_buffer_from_pipeline()` attach the audio buffer and then start recording, but they've drifted apart over time.

The difference is that `_attach_audio_buffer_from_pipeline()` handles the case where the `AudioBufferProcessor` changes by starting recording on the new processor directly. `attach_to_task()` doesn't do that, it always goes through `_ensure_audio_buffer_recording()`, which is where the stale `_abp_is_recording` flag can prevent recording from starting.

Since `attach_to_task()` is the code that's actually used, I'd suggest reusing `_attach_audio_buffer_from_pipeline()` (or merging the two into a single implementation) so this logic only lives in one place and doesn't get out of sync again.

Comment (shivamgcodes) [reply src/noveum_trace/integrations/pipecat/pipecat_observer.py]: Good catch — it turned out to be worse than drift: `_attach_audio_buffer_from_pipeline` has been dead code since the sync/async split in 284c531 ("made sync pipecat observer for dograh"). That commit moved the walk/handler registration into `_attach_audio_buffer_handler_sync` and the recording start into `_ensure_audio_buffer_recording`, but left the old combined method behind with no production caller (only tests referenced it; dograh uses `attach_to_task_sync` and starts ABP recording itself).

Removed the dead method in e982061 and retargeted the tests at the real `attach_to_task` path, so the attach-then-start logic now lives in one place. Re the stale `_abp_is_recording` concern: `_ensure_audio_buffer_recording` prefers the processor's own `_recording` attribute when present (real Pipecat ABPs have it), so a fresh processor still gets `start_recording()` called correctly — the internal flag is only a fallback for ABPs that don't expose `_recording`.

**Review by Vidushi2709 — APPROVED** [review]
